astropy/io/fits/hdu/compressed.py

```diff
@@ -1559,10 +1613,14 @@ def header(self):
             image_header.set('DATASUM', self._header['ZDATASUM'],
                              self._header.comments['ZDATASUM'])
 
+        # Clean up any possible doubled EXTNAME keywords that use
+        # the default.
+        self._remove_unnecessary_default_extnames(image_header)
```

Comment:
I think we could do that before copying `_header` (`image_header = self._header.copy()`), this way both `._header` and `_image_header` would be cleaned and calling `_remove_unnecessary_default_extnames` in the name setter would no more be necessary. I tried locally and it seems to work.

Reply:
I did originally do that but then I saw some comments about `_header` sometimes being read-only (depending on how the header is mapped from a file) and decided that it was safer to defer updating `_header` until the user actively asks for the header to be changed.

Reply:
Hmm, I don't think a `Header` could be read-only, but maybe I missed something.

Reply:
I've moved it to before the copy.